### PR TITLE
Feature/sn

### PR DESCRIPTION
**Pull request by hroest**
this should address the #1163

### CODE REVIEW

**Comment by hroest** [comment]
travis runs through but maybe @hendrikweisser can comment on this since it also affects the FeatureFinderIdentification 

basically it changes how the signal to noise score is computed: it is now computed from the position closest to the apex instead of whatever position `std::lower_bound` gave. This makes more sense to me but it changes the output values

**Comment by hendrikweisser** [comment]
Hannes, thanks for the heads-up. Your changes look okay to me. But I wonder: If `MSSpectrum::MZBegin` doesn't work the same on all platforms, shouldn't that be fixed, too?

**Comment by hroest** [comment]
@hendrikweisser we are looking into this but it really seems like `std::lower_bound` might be the issue since `MSSpectrum::MZBegin` is just a wrapper around that. I think the main problem is if you give it *identical* double values (e.g. 101 in this case). Still, I dont really understand the problem and to me it seems that from far away, Ubuntu 14.04 is doing the right thing as the specs are concerned and the other systems are actually wrong

**Comment by hroest** [comment]
Note that we can either merge this and improve the algorithm *and* fix the tests or we can merge #1191 and just fix the test in a really dirty way :-) since Hendrik is in favor of this and I am as well and it will improve the results I vote for this
@hendrikweisser I was a bit worried since the changes in the scores of the FFId were quite substantial

File: src/tests/topp/FeatureFinderIdentification_1_output.featureXML [hunk 1]
```diff
@@ -4275,10 +4275,10 @@
 			<userParam type="float" name="var_norm_rt_score" value="0.00189903781782697"/>
 			<userParam type="float" name="var_intensity_score" value="0.449663067292516"/>
 			<userParam type="float" name="nr_peaks" value="3"/>
-			<userParam type="float" name="sn_ratio" value="1.18492061027865"/>
-			<userParam type="float" name="var_log_sn_score" value="0.169675776797353"/>
+			<userParam type="float" name="sn_ratio" value="21.7975863988899"/>
```
Comment (hendrikweisser) [inline src/tests/topp/FeatureFinderIdentification_1_output.featureXML]: Okay, this is indeed a huge change (signal-to-noise ratio from 1.2 to 21.8). Any idea why that happens?

**Comment by hendrikweisser** [comment]
> I was a bit worried since the changes in the scores of the FFId were quite substantial

Some of them are, yes. But a) your change seems to be an improvement and b) I'm reworking the algorithm - and especially the scoring - anyway now.

**Comment by hroest** [comment]
yes, it would however still be nice to figure out why the changes are so
dramatic here and whether thre is an issue with my code or the test or
neither (and thats just how it is)

**Comment by hroest** [comment]
I further investigated why there is such a huge jump in score for one of the peptides 
```
1022: scoring feature RT: 2005.24 MZ: 0 INT: 406078 == LAADDFR/2 [ expected RT 0.502879 / 2003.34 ] with 3 nr transitions and nr chromats 3
1022:  get iterator here POS: 2007.43 INT: 22630.6 at 2007.43 --> for RT 2005.24
1022:    return 3.55476;
1022:    (for next) 4.76331 ; for prev: 3.85266
1022:  new way would return 3.85266 at POS: 2004.77 INT: 24527

1022:  get iterator here POS: 2007.43 INT: 0 at 2007.43 --> for RT 2005.24
1022:    return 0;
1022:    (for next) 0 ; for prev: 61.5401
1022:  new way would return 61.5401 at POS: 2004.77 INT: 4520.38

1022:  get iterator here POS: 2007.43 INT: 0 at 2007.43 --> for RT 2005.24
1022:    return 0;
1022:    (for next) 0 ; for prev: 0
1022:  new way would return 0 at POS: 2004.77 INT: 0
1022:  add sn_ration 1.18492

```

well it really seems that there is a jump in the SignalToNoiseEstimatorMedian value when evaluating at position 2004.77 instead of 2007.43 and the issue is that the former has an intensity of 4520 and the latter one of 0. Since we are most interested in the value at position 2005.24 here I would still argue that taking 2004 is better than 2007 since it is probably closer to the true value. This can also be seen from the image beow where the last data point is seen at RT=2004

@hendrikweisser are these input data centroided or how do you get a jump from 4500 to zero in just one scan? Otherwise I have not really seen such non-smooth data before, at least not really in SWATH-MS

data

**Comment by hroest** [comment]
ok, I am now convinced that this should be the way to go and that the strange behavior we have seen in hendriks tool are actually due to the underlying data and the algorithm seems to perform correctly in my eyes (and even improved). 

If @hendrikweisser agrees, I would opt for a merge

**Comment by hroest** [comment]
@hendrikweisser this might also point to a more general problem in the FFId since the S/N score is basically useless with highly reduced (centroided) data where you have mostly zeros and then a small stretch of signal where of course the S/N is high

**Comment by timosachsenberg** [comment]
+1 from me

**Comment by hendrikweisser** [comment]
@hroest: Thank you for looking into the large changes in the test results. :+1: for merging from me, too.

> are these input data centroided or how do you get a jump from 4500 to zero in just one scan? Otherwise I have not really seen such non-smooth data before, at least not really in SWATH-MS

Yes, the data is centroided. The centroiding step has been necessary - at least on the data I've initially tested on - to get (more or less) stable chromatograms after extraction. Without centroiding the intensities fluctuated too much. But perhaps I should do some more testing with different m/z window sizes to see if there's a "sweet spot" that works for extraction on raw MS1 data.

**Pull request merged**